### PR TITLE
Bring list-clouds help docs up-to-date

### PR DESCRIPTION
**Pull request by timClicks**
`juju list-clouds` changes its behaviour depending on whether there are current controllers. This commit makes this explicit.

Microk8s has not been previously mentioned.

The current text pre-supposes an Ubuntu-based Juju client. The help has been updated to include situations where the client does may have a local LXD instance. 

New `juju help clouds` output:

```
Usage: juju clouds [options]

Summary:
Lists all clouds available to Juju.

Global Options:
--debug  (= false)
    equivalent to --show-log --logging-config=<root>=DEBUG
-h, --help  (= false)
    Show help on a command or other topic.
--logging-config (= "")
    specify log levels for modules
--quiet  (= false)
    show no informational output
--show-log  (= false)
    if set, write the log file to stderr
--verbose  (= false)
    show more verbose output

Command Options:
-B, --no-browser-login  (= false)
    Do not use web browser for authentication
-c, --controller (= "")
    Controller to operate in
--format  (= tabular)
    Specify output format (json|tabular|yaml)
--local  (= false)
    Local operation only; controller not affected
-o, --output (= "")
    Specify an output file

Details:
Display the fundamental properties for each cloud known to the current Juju client:
name, number of regions, default region, type, and description

The default behaviour is to show clouds available on the current controller.
Another controller can specified using the --controller option. When no controllers
are available, --local is implied.

If --local is specified, the public clouds known to Juju out of the box are displayed,
along with any which have been added with `add-cloud --local`. These clouds can be
used to bootstrap a controller.

Clouds may be listed that are co-hosted with the Juju client.  When the LXD hypervisor
is detected, the 'localhost' cloud is made available.  When a microk8s installation is
detected, the 'microk8s' cloud is displayed.

This command's default output format is 'tabular'. Use 'json' and 'yaml' for
machine-readable output.

Cloud metadata sometimes changes, e.g. providers add regions. Use the `update-clouds`
command to update the current Juju client.

Use the `add-cloud` command to add a private cloud to the list of clouds known to the
current Juju client.

Use the `regions` command to list a cloud's regions.

Use the `show-cloud` command to get more detail, such as regions and endpoints.

Further reading:
 
    Documentation:   https://docs.jujucharms.com/stable/clouds
    microk8s:        https://microk8s.io/
    LXD hypervisor:  https://linuxcontainers.org/lxd/

Examples:

    juju clouds
    juju clouds --format yaml
    juju clouds --controller mycontroller
    juju clouds --local

See also:
    add-cloud
    credentials
    controllers
    regions
    set-default-credential
    set-default-region
    show-cloud
    update-clouds

Aliases: list-clouds
```

## QA steps

* mv ~/.local/share/juju/controllers.yaml ~/.local/share/juju/controllersx.yaml`
* `juju list-clouds`

> ```
> There are no controllers running.
> You can bootstrap a new controller using one of these clouds:
> Cloud                          Regions  Default          Type        Description
> aws                                 15  us-east-1        ec2         Amazon Web Services
> aws-china                            2  cn-north-1       ec2         Amazon China
> aws-gov                              1  us-gov-west-1    ec2         Amazon (USA Government)
> azure                               27  centralus        azure       Microsoft Azure
> azure-china                          2  chinaeast        azure       Microsoft Azure China
> cloudsigma                          12  dub              cloudsigma  CloudSigma Cloud
> google                              18  us-east1         gce         Google Cloud Platform
> joyent                               6  us-east-1        joyent      Joyent Cloud
> oracle                               4  us-phoenix-1     oci         Oracle Cloud Infrastructure
> oracle-classic                       5  uscom-central-1  oracle      Oracle Cloud Infrastructure Classic
> rackspace                            6  dfw              rackspace   Rackspace Cloud
> localhost                            1  localhost        lxd         LXD Container Hypervisor
> microk8s                             1  localhost        k8s         A Kubernetes Cluster
> ...
> ```

## Documentation changes

N/A

## Bug reference

* https://bugs.launchpad.net/juju/+bug/1830983

### CODE REVIEW

File: cmd/juju/cloud/list.go [hunk 1]
```diff
@@ -166,11 +167,12 @@ func (c *listCloudsCommand) Run(ctxt *cmd.Context) error {
 		}
 		output = clouds
 	default:
```
Comment (timClicks) [inline cmd/juju/cloud/list.go]: Simplifies some of the control flow, imo, but is functionally the same.

**Comment by timClicks** [comment]
[juju list-clouds should default to local if there are no controllers](https://trello.com/c/1XZj0wmB/780-juju-list-clouds-should-default-to-local-if-there-are-no-controllers)

File: cmd/juju/cloud/list.go [hunk 2]
```diff
@@ -166,11 +167,12 @@ func (c *listCloudsCommand) Run(ctxt *cmd.Context) error {
...
-		if c.controllerName == "" && !c.Local {
-			ctxt.Infof(
-				"There are no controllers running.\nYou can bootstrap a new controller using one of these clouds:\n")
-		}
-		if c.controllerName != "" {
+		if c.controllerName == "" {
```
Comment (wallyworld) [inline cmd/juju/cloud/list.go]: -1 on this change as it introduces extra indentation
The original logic was easy enough to read

File: cmd/juju/cloud/list.go [hunk 3]
```diff
@@ -38,23 +38,40 @@ type listCloudsCommand struct {
...
+	"In addition to these public clouds, clouds may be listed that are co-hosted with the\n" +
+	"Juju client.  When the LXD hypervisor is detected, the 'localhost' cloud is made\n" +
+	"available.  When a microk8s installation is detected, the 'microk8s' cloud \n" +
+	"is displayed.\n" +
+	"\n" +
+	"This command's default output format is 'tabular'. Use 'json' and 'yaml' for\n" +
+	"machine-readable output that also provides the endpoint URLs, authentication\n" +
+	"parameters, and any cloud-specific configuration that has been set.\n" +
+	"\n" +
+	"Cloud metadata sometimes changes, e.g. providers add regions and instance types. Use \n" +
```
Comment (wallyworld) [inline cmd/juju/cloud/list.go]: instance types is not part of cloud metadata

File: cmd/juju/cloud/list.go [hunk 4]
```diff
@@ -38,23 +38,40 @@ type listCloudsCommand struct {
...
-	"or another controller specified using --controller.\n" +
-	"\nIf --local is specified, the output shows public clouds known to Juju out of the box;\n" +
-	"these can be used to bootstrap a controller. Clouds may change between Juju versions.\n" +
-	"In addition to these public clouds, the 'localhost' cloud (local LXD) is also listed.\n" +
-	"If you supply a controller name the clouds known on the controller will be displayed.\n" +
-	"\nThis command's default output format is 'tabular'.\n" +
-	"\nCloud metadata sometimes changes, e.g. AWS adds a new region. Use the\n" +
-	"`update-clouds` command to update the current Juju client accordingly.\n" +
-	"\nUse the `add-cloud` command to add a private cloud to the list of\n" +
+	"Display the clouds known to the current Juju client, as well as the number of region\n" +
```
Comment (wallyworld) [inline cmd/juju/cloud/list.go]: regions

File: cmd/juju/cloud/list.go [hunk 5]
```diff
@@ -38,23 +38,40 @@ type listCloudsCommand struct {
...
+	"that each cloud supports, their default regions, their types, and short descriptions.\n" +
```
Comment (wallyworld) [inline cmd/juju/cloud/list.go]: I preferred the original opening sentence to be honest. More concise and meaningful in that it mentioned what was displayed - fundamental cloud properties.

File: cmd/juju/cloud/list.go [hunk 6]
```diff
@@ -38,23 +38,40 @@ type listCloudsCommand struct {
...
+	"\n" +
+	"The default behaviour is to show clouds available on the current controller.\n" +
+	"Another controller can specified using the --controller option. When no controllers\n" +
+	"are available, --local is implied.\n" +
+	"\n" +
+	"If --local is specified, the output shows public clouds known to Juju out of the box;\n" +
+	"these can be used to bootstrap a controller.\n" +
```
Comment (pmatulis) [inline cmd/juju/cloud/list.go]: create a controller

File: cmd/juju/cloud/list.go (hunk 5 above)
Comment (timClicks) [reply cmd/juju/cloud/list.go]: Have reverted, but expanded the line-width to what Chris had been using.

File: cmd/juju/cloud/list.go (hunk 6 above)
Comment (timClicks) [reply cmd/juju/cloud/list.go]: Good point. Have changed.

**Comment by timClicks** [comment]
$$merge$$

**Comment by timClicks** [comment]
$$merge$$

**Pull request merged**